File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    // Global
    vector<vector<int>> ans;

    void solve(vector<int> &candi, int target, vector<int> &temp, int index, int sum){
        if (sum == target) {
            ans.push_back(temp);
            return ;
        }

        if (sum > target || index == candi.size()) return ;

        temp.push_back(candi[index]);
        solve(candi, target, temp, index, sum+candi[index]);
        temp.pop_back();
        solve(candi, target, temp, index+1, sum);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        sort(candidates.begin(), candidates.end());

        solve(candidates, target, temp, 0, 0);

        return ans;
    }
};

File: 39-combination-sum/combination-sum.go
func solve(candidates []int, target, sum, idx int, temp *[]int, ans *[][]int) {
	if sum == target {
        sol := make([]int, len(*temp))
        copy(sol, *temp)
		*ans = append(*ans, sol)
		return
	}
	if sum > target || idx >= len(candidates) {
		return
	}
	*temp = append(*temp, candidates[idx])
	solve(candidates, target, sum+candidates[idx], idx, temp, ans)
	*temp = (*temp)[:len(*temp)-1]
	solve(candidates, target, sum, idx+1, temp, ans)
}

func combinationSum(candidates []int, target int) [][]int {
	var temp []int
	var ans [][]int
	solve(candidates, target, 0, 0, &temp, &ans)
	return ans
}